compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs

```diff
@@ -319,6 +325,18 @@ fn add_unused_functions<'ll, 'tcx>(cx: &CodegenCx<'ll, 'tcx>) {
             continue;
         }
 
+        // If this function hasn't been instrumented with code coverage for some reason, then skip
+        // it so we don't try to instantiate the function and related statics improperly (#93054).
+        let coverageinfo =
+            tcx.coverageinfo(InstanceDef::Item(WithOptConstParam::unknown(non_codegenned_def_id)));
+        if coverageinfo.num_counters == 1 && coverageinfo.num_expressions == 0 {
```

Comment:
@richkadel I tried to use your suggestion of `!tcx.covered_code_regions(non_codegenned_def_id).is_empty()` here but that doesn't actually catch the functions which triggers the assert. I had to use this check instead. Does that seem reasonable to you?

Reply:
My bad. I forgot how complex it is to compute `counter_regions` passed into `write_coverage_mapping`.

In that case, I think your original solution is simpler. Sorry. 

Can I suggest, though, that you change it to (abbreviated):

```
if coverage_mapping_buffer.is_empty() {
  if function_coverage.is_used()) {
    bug!("used functions should have coverage, but {} had none", function_name);
  } else {
    debug!("unused function {} had no coverage", function_name);
  }
}

Reply:
Ok, done!